mavros_extras/src/plugins/esc_status.cpp

```diff
@@ -0,0 +1,148 @@
+/**
+ * @brief ESC status plugin
+ * @file esc_status.cpp
+ * @author Ricardo Marques <marques.ricardo17@gmail.com>
+ *
+ * @addtogroup plugin
+ * @{
+ */
+/*
+ * Copyright 2020 Ricardo Marques <marques.ricardo17@gmail.com>.
+ *
+ * This file is part of the mavros package and subject to the license terms
+ * in the top-level LICENSE file of the mavros repository.
+ * https://github.com/mavlink/mavros/tree/master/LICENSE.md
+ */
+
+#include <mavros/mavros_plugin.h>
+#include <mavros_msgs/ESCInfo.h>
+#include <mavros_msgs/ESCStatus.h>
+
+namespace mavros
+{
+namespace extra_plugins
+{
+/**
+ * @brief ESC status plugin
+ */
+class ESCStatusPlugin : public plugin::PluginBase
+{
+public:
+	ESCStatusPlugin() : PluginBase(),
+		_max_esc_info_index(0),
+		_max_esc_status_index(0),
+		nh("~")
+	{}
+
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+
+		esc_info_pub = nh.advertise<mavros_msgs::ESCInfo>("esc_info", 10);
+		esc_status_pub = nh.advertise<mavros_msgs::ESCStatus>("esc_status", 10);
+
+		enable_connection_cb();
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return {
+			       make_handler(&ESCStatusPlugin::handle_esc_info),
+			       make_handler(&ESCStatusPlugin::handle_esc_status),
+		};
+	}
+
+private:
+	using lock_guard = std::lock_guard<std::mutex>;
+	std::mutex mutex;
+
+	ros::NodeHandle nh;
+
+	ros::Publisher esc_info_pub;
+	ros::Publisher esc_status_pub;
+	mavros_msgs::ESCInfo _esc_info;
+	mavros_msgs::ESCStatus _esc_status;
+	uint8_t _max_esc_info_index;
+	uint8_t _max_esc_status_index;
+	const uint8_t batch_size = 4;
+
+	void handle_esc_info(const mavlink::mavlink_message_t *msg, mavlink::common::msg::ESC_INFO &esc_info)
+	{
+		lock_guard lock(mutex);
+
+		_esc_info.header.stamp = m_uas->synchronise_stamp(esc_info.time_usec);
+
+		uint8_t esc_index = esc_info.index;
+
+		_esc_info.counter = esc_info.counter;
+		_esc_info.count = esc_info.count;
+		_esc_info.connection_type = esc_info.connection_type;
+		_esc_info.info = esc_info.info;
+
+		size_t required_size = _esc_info.count;
+		if (_esc_info.esc_info.size() < required_size)
+		{
+			_esc_info.esc_info.resize(_esc_info.count);
+		}
+
+		for (int i = 0; i < batch_size; i++)
+		{
+			_esc_info.esc_info[esc_index + i].header =  _esc_info.header;
+			_esc_info.esc_info[esc_index + i].failure_flags = esc_info.failure_flags[i];
+			_esc_info.esc_info[esc_index + i].error_count = esc_info.error_count[i];
+			_esc_info.esc_info[esc_index + i].temperature = esc_info.temperature[i];
```

Comment:
Error in index calculation: you should: `batch_size * esc_index + i`.

```
auto &e = _esc_info.esc_info[batch_size * esc_index + i];
e.header = _esc_info.header
...
```

Reply:
esc_index is either 0, 4, 8 etc

MAVLink message:

`<field type="uint8_t" name="index">Index of the first ESC in this message. minValue = 0, maxValue = 60, increment = 4.</field>`

So I think like this it's correct no? Let me know if further changes are needed here.

Reply:
Oh, again i am not reading carefully. But that also based on other messages, where index increment is 1.

Reply:
no worries @vooon I understand and it's why I'm here :)